In [109]:
import numpy as np
from Bio import SeqIO
from Bio import Entrez
import string
import time
import matplotlib as plt
import pandas as pd

In [86]:
# generating files of various 0-content
#zeros_100 = np.random.choice([0, 1], size=800000000, replace=True, p=[1.0, 0.0]) 
zeros_90 = np.random.choice([0, 1], size=800000000, replace=True, p=[0.9, 0.1])
#zeros_80 = np.random.choice([0, 1], size=800000000, replace=True, p=[0.8, 0.2])
#zeros_70 = np.random.choice([0, 1], size=800000000, replace=True, p=[0.7, 0.3])
#zeros_60 = np.random.choice([0, 1], size=800000000, replace=True, p=[0.6, 0.4])
#zeros_50 = np.random.choice([0, 1], size=800000000, replace=True, p=[0.5, 0.5])

#zeros_100 = np.packbits(zeros_100)
zeros_90 = np.packbits(zeros_90)
#zeros_80 = np.packbits(zeros_80)
#zeros_70 = np.packbits(zeros_70)
#zeros_60 = np.packbits(zeros_60)
#zeros_50 = np.packbits(zeros_50)

#open('zeros_100', 'wb').write(zeros_100)
open('zeros_90', 'wb').write(zeros_90)
#open('zeros_80', 'wb').write(zeros_80)
#open('zeros_70', 'wb').write(zeros_70)
#open('zeros_60', 'wb').write(zeros_60)
#open('zeros_50', 'wb').write(zeros_50)

100000000

In [60]:
# generating random dna sequence and protein files
random_dna_seq_nums = np.random.choice([1, 2, 3, 4], size=100000000, replace=True, p=[.25, .25, .25, .25])
random_dna_seq_array = []
for elem in random_dna_seq_nums:
    if elem == 1:
        random_dna_seq_array.append('A')
    if elem == 2:
        random_dna_seq_array.append('G')
    if elem == 3:
        random_dna_seq_array.append('T')
    if elem == 4:
        random_dna_seq_array.append('C')
random_dna_seq = ''.join(random_dna_seq_array)

aa_dict = {1:'A', 2:'R', 3:'N', 4:'D', 5:'C', 6:'Q', 7:'E', 8:'G', 9:'H', 10:'I', 11:'M', 12:'L', 13:'K', 14:'P', 15:'F', 16:'S', 17:'T', 18:'W', 19:'Y', 20:'V'}
random_prot_seq_nums = np.random.choice(range(1,21), size=100000000, replace=True)
random_prot_seq_array = []
for elem in random_prot_seq_nums:
    random_prot_seq_array.append(aa_dict[elem])
random_prot_seq = ''.join(random_prot_seq_array)

open('dna_seq.fa', 'w').write(''.join(random_dna_seq_array))
open('prot_seq.fa', 'w').write(''.join(random_prot_seq_array))



100000000

## Time Trial Data: GZIP

### Terminal Commands
##### > time gzip -c zeros_100 > zeros_100_gzip.gz
##### > time gzip -c zeros_90 > zeros_90_gzip.gz
##### > time gzip -c zeros_80 > zeros_80_gzip.gz
##### > time gzip -c zeros_70 > zeros_70_gzip.gz
##### > time gzip -c zeros_60 > zeros_60_gzip.gz
##### > time gzip -c zeros_50 > zeros_50_gzip.gz

In [105]:
data = {}
data['info'] = ['real time (s)', 'user time (s)', 'sys time (s)', 'input size (MB)', 'output size (MB)']
data['zeros_100_gzip'] = [7.948, 0.554, 0.036, 100, .097081]
data['zeros_90_gzip'] = [30.788, 25.673, 0.076, 100, 56.012]
data['zeros_80_gzip'] = [16.872, 16.500, 0.112, 100, 77.402]
data['zeros_70_gzip'] = [6.788, 6.744, 0.044, 100, 89.280]
data['zeros_60_gzip'] = [4.217, 4.149, 0.068, 100, 97.674]
data['zeros_50_gzip'] = [3.588, 3.358, 0.056, 100, 100.015]
data['DNA Sequence'] = [23.071, 22.953, 0.056, 100, 29.222]
data['Protein Sequence'] = [4.247, 4.190, 0.056, 100, 60.559]

df = pd.DataFrame(data)
df

info  zeros_100_gzip  zeros_90_gzip  zeros_80_gzip  \
0     real time (s)        7.948000         30.788         16.872   
1     user time (s)        0.554000         25.673         16.500   
2      sys time (s)        0.036000          0.076          0.112   
3   input size (MB)      100.000000        100.000        100.000   
4  output size (MB)        0.097081         56.012         77.402   

   zeros_70_gzip  zeros_60_gzip  zeros_50_gzip  DNA Sequence  Protein Sequence  
0          6.788          4.217          3.588        23.071             4.247  
1          6.744          4.149          3.358        22.953             4.190  
2          0.044          0.068          0.056         0.056             0.056  
3        100.000        100.000        100.000       100.000           100.000  
4         89.280         97.674        100.015        29.222            60.559

## Time Trial Data: BZIP2

### Terminal Commands
##### > time bzip -c zeros_100 > zeros_100_bzip.bz2 
##### (and for all the other files replace 100 with respective 0-content percentage)

In [168]:
data2 = {}
data2['info'] = ['real time (s)', 'user time (s)', 'sys time (s)', 'input size (MB)', 'output size (MB)']
data2['zeros_100_bzip'] = [0.861, 0.823, 0.036, 100, .000113]
data2['zeros_90_bzip'] = [9.488, 9.407, 0.080, 100, 58.321]
data2['zeros_80_bzip'] = [9.955, 9.846, 0.108, 100, 82.634]
data2['zeros_70_bzip'] = [10.746, 10.666, 0.080, 100, 95.138]
data2['zeros_60_bzip'] = [11.912, 11.842, 0.068, 100, 100.050]
data2['zeros_50_bzip'] = [12.687, 12.602, 0.084, 100, 100.442]
data2['DNA Sequence'] = [9.306, 9.254, 0.052, 100, 27.335]
data2['Protein Sequence'] = [9.372, 9.300, 0.072, 100, 55.260]
df = pd.DataFrame(data2)
df

info  zeros_100_bzip  zeros_90_bzip  zeros_80_bzip  \
0     real time (s)        0.861000          9.488          9.955   
1     user time (s)        0.823000          9.407          9.846   
2      sys time (s)        0.036000          0.080          0.108   
3   input size (MB)      100.000000        100.000        100.000   
4  output size (MB)        0.000113         58.321         82.634   

   zeros_70_bzip  zeros_60_bzip  zeros_50_bzip  DNA Sequence  Protein Sequence  
0         10.746         11.912         12.687         9.306             9.372  
1         10.666         11.842         12.602         9.254             9.300  
2          0.080          0.068          0.084         0.052             0.072  
3        100.000        100.000        100.000       100.000           100.000  
4         95.138        100.050        100.442        27.335            55.260

## Time Trial Data: PBZIP2

### Terminal Commands
##### > time pbzip2 -c zeros_50 > zeros_50_pbzip.bz2 
##### (run 5 times for each zero percentage)
##### > ls -l -t *.bz2 to get more precise file size data

In [169]:
data3 = {}
data3['info'] = ['real time (s)', 'user time (s)', 'sys time (s)', 'input size (MB)', 'output size (MB)']
data3['zeros_100_pbzip'] = [0.142, 1.776, 0.042, 100, .005375]
data3['zeros_90_pbzip'] = [1.017, 14.953, 0.542, 100, 58.343]
data3['zeros_80_pbzip'] = [1.182, 17.439, 0.631, 100, 82.649]
data3['zeros_70_pbzip'] = [1.413, 21.123, .633, 100, 95.143]
data3['zeros_60_pbzip'] = [1.680, 25.348, 0.598, 100, 100.057]
data3['zeros_50_pbzip'] = [1.821, 27.539, .643, 100, 100.451]
data3['DNA_Sequence'] = [1.052, 14.148, 0.536, 100, 27.342]
data3['Protein Sequence'] = [1.119, 14.573, .607, 100, 55.259]

df = pd.DataFrame(data3)
df

info  zeros_100_pbzip  zeros_90_pbzip  zeros_80_pbzip  \
0     real time (s)         0.142000           1.017           1.182   
1     user time (s)         1.776000          14.953          17.439   
2      sys time (s)         0.042000           0.542           0.631   
3   input size (MB)       100.000000         100.000         100.000   
4  output size (MB)         0.005375          58.343          82.649   

   zeros_70_pbzip  zeros_60_pbzip  zeros_50_pbzip  DNA_Sequence  \
0           1.413           1.680           1.821         1.052   
1          21.123          25.348          27.539        14.148   
2           0.633           0.598           0.643         0.536   
3         100.000         100.000         100.000       100.000   
4          95.143         100.057         100.451        27.342   

   Protein Sequence  
0             1.119  
1            14.573  
2             0.607  
3           100.000  
4            55.259

Gzip is the best compression algorithm for all files of zeros and ones. Pbzip is the best compression algorithm (by a very small margin) for the protein sequence, and bzip is the best compression algorithm for the random DNA sequence (by a very small margin). Pbzip2 is the fastest algorithm. It is an optimized algorithm that differs from bzip2 by its ability to do multi-threading and therefore gave almost linear speed improvements. Therefore, we expect pbzip2 to be fastest across the board (http://www.softpanorama.org/Tools/Archivers/bzip2_and_pbzip2.shtml). 

The level of compression decreases as the percentage of zeros decreases. This is because with block codes, or run-length encoding, the compression is optimized with longs runs of zeros (which will be more frequent in the files with higher zero content). In the 50% zero file, the average run may be less than 2.

The minimum number of bits to encode a single dna base is 2. The minimum number of bits to store an amino acid is 4.3 because there are 20 amino acids, and log(20) = 4.3. The gzip file on random dna took (29.22MB * 100,000 * 8bits) / 100,000,000 nucleotides =  2.34 bits per dna base. The bzip2 file on random dna took (27.342 * 100,000 * 8bits) / 100,000,000 nucleotides = 2.19 bits per dna base. The gzip file on random aa sequence took (55.26MB * 100,000 * 8bits) / 100,000,000 amino acids = 4.42 bits per amino acid. The pbzip on random aa sequence took also 4.42 bits per aa with the same calculation. 

Gzip and bzip are not performing optimally on dna and proteins. When dna sequences are 100 million base pairs long, compression ineffiency of even .3 bits can make a huge difference. 

## Real Data

In [160]:
# extracting 10 HIV isolates from genbank based on their accession numbers and turning them into multi-fasta string
Entrez.email = 'sylviaillouz@berkeley.edu'
handle = Entrez.esearch(db ='nucleotide', term =  'gp120[All Fields]', sort ='relevance', idtype ='acc')
i = Entrez.read(handle)['IdList'][0]
fasta_seq = ""

hiv_id_nums = ['AF236860.1', 'AF236859.1', 'AF230868.1', 'FN557331.1', 'Y14419.1', 'Y14411.1', 'Y14412.1', 'Y14414.1', 'Y14413.1', 'Y14416.1']
for id_num in hiv_id_nums:
    handle = Entrez.efetch(db ='nucleotide', id = id_num, rettype = 'gb', retmode = 'text', retmax = 1)
    for record in SeqIO.parse(handle, 'genbank'):
        fasta_seq += "> " + str(record.id) + " | " + str(record.description) + "\n" + str(record.seq) + "\n\n"
        
print(fasta_seq)

> AF236860.1 | HIV-1 strain TH90129_1 from Thailand gp120 (gp120) gene, partial cds
GTTCCTGTGTGGAAAGATGCAGAGACCACCTTATTTTGTGCATCAGATGCCAAAGCACATGAGACAGAAGTGCACAATGTCTGGGCCACACATGCCTGTGTACCCACAGACCCCAACCCACAAGAAATACAACTGAAAAATGTAACAGAGAATTTTAACATGTGGAAAAATAACATGGTAGAGCAGATGCAGGAGGATGTAATCAGTTTATGGGATCAAAGTCTAAAGCCATGTGTAAAGTTAACTCCTCTCTGCGTTACTTTAAATTGTACCGATGCTACTTTGACCAATAGCACTTACATAACCAATGTCTCTAAGATAATAGGAGATATAACAGAGGAAGTAAGAAACTGTTCTTTTAATATGACCACAGAACTAAGAGATAAGAAGCAGAAGGTCCATGCACTTTTTTTATAAGCTTGATATAGTAGAAATTGAAAAGAATAGGAATGAGTATAGGTTAATAAATTGTAATACTTCGGTCATTAAGCAGGCTTGTCCAAAGATATCCTTTGATCCAATTCCTATACATTATTGTACTCCAGCTGGTTATGCGATTTTAAAGTGTAATGATAAGAATTTCAATGGGACAGGGCCATGTAAAAATGTCAGCTCAGTACAATGCACACATGGAATTAAGCCAGTGGTATCAACTCAATTGCTGTTAAATGGCAGTCTAGCAGAAGAAGAGATAATAATCAGATCTGAAAATCTCACAAACAATGCCAAAACCATAATAGTGCACCTTAATAAATCTGTAGAAATCAATTGTACCAGACCCTTCAACAATACAAGAACAAGTATAACTATAGGACCAGGACAAATGTTCTATAGAACAGGAGAGATAATAGGAGATATAACAAAAGCATATTGTGAGATTAATGGAACAAAATGGAATGAAACTTTAAAACAGGTA

In [167]:
open('HIV_isolates.fa', 'w').write(fasta_seq)

8786

The non random files should have a better compression ratios because in real DNA, there are many multi-nucleotide runs of the same dna base (A, T, G, C) that approach a more optimized run-length encoding algorithm like gzip. 

### Terminal Commands:
##### > time pbzip -c HIV_isolates.fa > HIV_pbzip.bz2
#### > ls -l -t *.bz2
 
##### > time gzip -c HIV-isolates.fa > HIV_gzip.gz
##### > ls -l - t *.gz

### Outputs:
####  compression ratio pbzip: 2495 kB / 8786 kB = 0.284
#### compression ration gzip: 2516 kB / 8786 kB = .286

#### compression ratio random pbzip: 100 MB / 100 MB = 1
#### compressio ratio random gzip: 100 MB / 100 MB = 1
This compression ratio for the random 50/50 0 and 1 files is much worse than the compression ratio for real genomic data.

### Estimation

For the binary micriscope images with completely random data, I would recommend gzip to compress. For the protein sequences, I would recommend pbzip, and pbzip for the re-sequencing of genomes data. With these compression schemes we should save roughly:
    
    (0.8)(0.286)(1000 terabytes from the dna seq data)
    
    (0.1)(.286)(1000 terabytes from the protein seq)
    
    (0.1)(0.85)(1000 terabytes from the random binary data)
    
    
   = Roughly 675 terabytes saved for each 1000
   
   67.5% space on average
   
   $337500 bonus
  
    